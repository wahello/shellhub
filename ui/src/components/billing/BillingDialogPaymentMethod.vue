<template>
  <fragment>
    <v-btn
      outlined
      data-test="show-btn"
      @click="show"
    >
      {{ typeOperation }}
    </v-btn>

    <v-dialog
      v-model="dialog"
      max-width="600"
      @click:outside="dialog=!dialog"
    >
      <v-card data-test="BillingDialogPaymentMethod-dialog">
        <v-card-title
          class="headline grey lighten-2 text-center"
          data-test="text-cardTitle"
        >
          {{ typeOperation | capitalizeFirstLetter }} payment method
        </v-card-title>

        <v-card-text class="mt-4 mb-3 pb-1">
          <div v-if="typeOperation === 'subscription'">
            Add payment method information below
          </div>

          <v-card class="mt-8">
            <v-col>
              <div ref="card" />
            </v-col>
          </v-card>

          <div
            ref="card-element-errors"
            class="card-errors mt-4"
            role="alert"
          />

          <v-spacer />

          <v-row class="mt-2">
            <v-spacer />
            <v-col
              md="auto"
              class="ml-auto"
            />
          </v-row>
        </v-card-text>

        <v-card-actions>
          <v-spacer />

          <v-btn
            text
            data-test="cancel-btn"
            @click="dialog=!dialog"
          >
            Close
          </v-btn>

          <v-btn
            text
            data-test="confirm-btn"
            :disabled="lockButton"
            @click="doAction()"
          >
            confirm
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

import capitalizeFirstLetter from '@/components/filter/string';

export default {
  name: 'BillingDialogPaymentMethod',

  filters: { capitalizeFirstLetter },

  props: {
    typeOperation: {
      type: String,
      default: 'subscription',
      validator: (value) => ['subscription', 'update'].includes(value),
    },
  },

  data() {
    return {
      dialog: false,
      card: null,
      elementError: null,
      elements: null,
      lockButton: false,
    };
  },

  computed: {
    requestWaiting() {
      return this.$store.getters['spinner/getStatus'];
    },
  },

  methods: {
    show() {
      this.dialog = !this.dialog;
      this.$nextTick(async () => {
        await this.mountStripeElements();
      });
    },

    displayError(e) {
      if (e.error) {
        this.elementError.textContent = e.error.message;
      } else {
        this.elementError.textContent = '';
      }
    },

    doAction() {
      this.lockButton = true;

      switch (this.typeOperation) {
      case 'subscription':
        this.subscriptionPaymentMethod();
        break;
      case 'update':
        this.updatePaymentMethod();
        break;
      default:
        this.lockButton = false;
      }
    },

    async mountStripeElements() {
      this.elements = this.$stripe.elements();
      this.card = this.elements.create('card');
      this.card.mount(this.$refs.card);
      this.elementError = this.$refs['card-element-errors'];
      this.card.on('change', (ev) => {
        this.displayError(ev);
      });
    },

    async subscriptionPaymentMethod() {
      const result = await this.$stripe.createPaymentMethod({
        type: 'card',
        card: this.card,
      });

      if (result.error) {
        this.displayError(result.error);
      } else {
        try {
          await this.$store.dispatch('billing/subscritionPaymentMethod', {
            payment_method_id: result.paymentMethod.id,
          });
          this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.subscription);
          this.$emit('update');
        } catch {
          this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.subscription);
        }
      }

      this.lockButton = false;
    },

    async updatePaymentMethod() {
      const result = await this.$stripe.createPaymentMethod({
        type: 'card',
        card: this.card,
      });

      if (result.error) {
        this.displayError(result.error);
      } else {
        try {
          await this.$store.dispatch('billing/updatePaymentMethod', {
            payment_id: result.paymentMethod.id,
          });
          this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.updateSubscription);
          this.$emit('update');
        } catch {
          this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.updateSubscription);
        }
      }

      this.lockButton = false;
    },
  },
};

</script>
