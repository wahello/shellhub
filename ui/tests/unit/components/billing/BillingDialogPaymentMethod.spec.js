import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import BillingDialogPaymentMethod from '@/components/billing/BillingDialogPaymentMethod';

describe('BillingDialogPaymentMethod', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);

  document.body.setAttribute('data-app', true);

  let wrapper;

  const typeOperation = 'subscription';
  const hasSpinner = false;

  const store = new Vuex.Store({
    namespaced: true,
    state: {
      hasSpinner,
    },
    getters: {
      'spinner/getStatus': (state) => state.hasSpinner,
    },
    actions: {
      'billing/subscritionPaymentMethod': () => {},
      'billing/updatePaymentMethod': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorAction': () => {},
    },
  });

  ///////
  // In this case, it's testing the button rendering.
  ///////

  describe('Button', () => {
    beforeEach(() => {
      wrapper = mount(BillingDialogPaymentMethod, {
        localVue,
        store,
        stubs: ['fragment'],
        propsData: { typeOperation },
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    /// //
    // Data and Props checking
    /// /

    it('Receive data in props', () => {
      expect(wrapper.vm.typeOperation).toBe(typeOperation);
    });
    it('Compare data with default value', () => {
      expect(wrapper.vm.dialog).toEqual(false);
      expect(wrapper.vm.card).toEqual(null);
      expect(wrapper.vm.elementError).toEqual(null);
      expect(wrapper.vm.elements).toEqual(null);
      expect(wrapper.vm.lockButton).toEqual(false);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="show-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="BillingDialogPaymentMethod-dialog"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="cancel-btn"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="confirm-btn"]').exists()).toBe(false);
    });
  });

  ///////
  // In this case, it's testing the subscription.
  ///////

  describe('Dialog Subscription', () => {
    beforeEach(() => {
      wrapper = mount(BillingDialogPaymentMethod, {
        localVue,
        store,
        stubs: ['fragment'],
        propsData: { typeOperation },
        vuetify,
      });

      wrapper.setData({ dialog: true });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    /// //
    // Data and Props checking
    /// /

    it('Receive data in props', () => {
      expect(wrapper.vm.typeOperation).toBe(typeOperation);
    });
    it('Compare data with default value', () => {
      expect(wrapper.vm.dialog).toEqual(true);
      expect(wrapper.vm.card).toEqual(null);
      expect(wrapper.vm.elementError).toEqual(null);
      expect(wrapper.vm.elements).toEqual(null);
      expect(wrapper.vm.lockButton).toEqual(false);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="show-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="BillingDialogPaymentMethod-dialog"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="cancel-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="confirm-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="text-cardTitle"]').text()).toEqual('Subscription payment method');
    });
  });

  ///////
  // In this case, it's testing the update subscription.
  ///////

  describe('Dialog Update', () => {
    const typeOperationUpdate = 'update';

    beforeEach(() => {
      wrapper = mount(BillingDialogPaymentMethod, {
        localVue,
        store,
        stubs: ['fragment'],
        propsData: { typeOperation: typeOperationUpdate },
        vuetify,
      });

      wrapper.setData({ dialog: true });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    /// //
    // Data and Props checking
    /// /

    it('Receive data in props', () => {
      expect(wrapper.vm.typeOperation).toBe(typeOperationUpdate);
    });
    it('Compare data with default value', () => {
      expect(wrapper.vm.dialog).toEqual(true);
      expect(wrapper.vm.card).toEqual(null);
      expect(wrapper.vm.elementError).toEqual(null);
      expect(wrapper.vm.elements).toEqual(null);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="show-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="BillingDialogPaymentMethod-dialog"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="cancel-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="confirm-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="text-cardTitle"]').text()).toEqual('Update payment method');
    });
  });
});
