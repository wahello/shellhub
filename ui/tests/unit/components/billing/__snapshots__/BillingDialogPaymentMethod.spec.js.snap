// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillingDialogPaymentMethod Button Renders the component 1`] = `
<fragment-stub><button type="button" class="v-btn v-btn--outlined theme--light v-size--default" data-test="show-btn"><span class="v-btn__content">
    subscription
  </span></button>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`BillingDialogPaymentMethod Dialog Subscription Renders the component 1`] = `
<fragment-stub><button type="button" class="v-btn v-btn--outlined theme--light v-size--default" data-test="show-btn"><span class="v-btn__content">
    subscription
  </span></button>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`BillingDialogPaymentMethod Dialog Update Renders the component 1`] = `
<fragment-stub><button type="button" class="v-btn v-btn--outlined theme--light v-size--default" data-test="show-btn"><span class="v-btn__content">
    update
  </span></button>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;
